cloudsmith/data_source_package.go

```diff
@@ -0,0 +1,185 @@
+package cloudsmith
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path"
+
+	"net/http"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+func dataSourcePackageRead(d *schema.ResourceData, m interface{}) error {
+	pc := m.(*providerConfig)
+	namespace := requiredString(d, "namespace")
+	repository := requiredString(d, "repository")
+	identifier := requiredString(d, "identifier")
+	download := d.Get("download").(bool)
+	outputPath := d.Get("output_path").(string)
+
+	req := pc.APIClient.PackagesApi.PackagesRead(pc.Auth, namespace, repository, identifier)
+	pkg, _, err := pc.APIClient.PackagesApi.PackagesReadExecute(req)
+	if err != nil {
+		return err
+	}
+
+	d.Set("cdn_url", pkg.GetCdnUrl())
+	d.Set("format", pkg.GetFormat())
+	d.Set("is_sync_awaiting", pkg.GetIsSyncAwaiting())
+	d.Set("is_sync_completed", pkg.GetIsSyncCompleted())
+	d.Set("is_sync_failed", pkg.GetIsSyncFailed())
+	d.Set("is_sync_in_flight", pkg.GetIsSyncInFlight())
+	d.Set("is_sync_in_progress", pkg.GetIsSyncInProgress())
+	d.Set("name", pkg.GetName())
+	d.Set("slug", pkg.GetSlug())
+	d.Set("slug_perm", pkg.GetSlugPerm())
+	d.Set("version", pkg.GetVersion())
+
+	d.SetId(fmt.Sprintf("%s_%s_%s", namespace, repository, pkg.GetSlugPerm()))
+
+	if download {
+		err := downloadPackage(pkg.GetCdnUrl(), outputPath, pc.GetAPIKey())
+		if err != nil {
+			return err
+		}
+	} else {
+		d.Set("output_path", pkg.GetCdnUrl())
+	}
```

Comment:
with the branches here we end up with different types of path for `output_path`, one is going to be the folder in which the file is downloaded, the other is going to be a full URL including the package/file name

wondering what we should do here, do we need a `download_dir`schema variable or something? then we can set `output_path` to `computed` and then just always set that to the right _full_ path... what do you think?